OpenCL/m24600-pure.cl

```diff
@@ -0,0 +1,442 @@
+
+    unpack64v (tmps, out, gid, i + 0, out[0]);
+    unpack64v (tmps, out, gid, i + 1, out[1]);
+    unpack64v (tmps, out, gid, i + 2, out[2]);
+    unpack64v (tmps, out, gid, i + 3, out[3]);
+    unpack64v (tmps, out, gid, i + 4, out[4]);
+    unpack64v (tmps, out, gid, i + 5, out[5]);
+    unpack64v (tmps, out, gid, i + 6, out[6]);
+    unpack64v (tmps, out, gid, i + 7, out[7]);
+  }
+}
+
+KERNEL_FQ void m24600_comp (KERN_ATTR_TMPS_ESALT (pbkdf2_sha512_tmp_t, pbkdf2_sha512_t))
+{
+  /**
+   * base
+   */
+
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  const u64 lid = get_local_id (0);
+
+  const u64 lsz = get_local_size(0);
+   /**
+   * aes shared
+   */
+
+  #ifdef REAL_SHM
+
+  LOCAL_VK u32 s_td0[256];
+  LOCAL_VK u32 s_td1[256];
+  LOCAL_VK u32 s_td2[256];
+  LOCAL_VK u32 s_td3[256];
+  LOCAL_VK u32 s_td4[256];
+
+  LOCAL_VK u32 s_te0[256];
+  LOCAL_VK u32 s_te1[256];
+  LOCAL_VK u32 s_te2[256];
+  LOCAL_VK u32 s_te3[256];
+  LOCAL_VK u32 s_te4[256];
+
+  for (u32 i = lid; i < 256; i += lsz)
+  {
+    s_td0[i] = td0[i];
+    s_td1[i] = td1[i];
+    s_td2[i] = td2[i];
+    s_td3[i] = td3[i];
+    s_td4[i] = td4[i];
+
+    s_te0[i] = te0[i];
+    s_te1[i] = te1[i];
+    s_te2[i] = te2[i];
+    s_te3[i] = te3[i];
+    s_te4[i] = te4[i];
+  }
+
+  SYNC_THREADS ();
+
+  #else
+
+  CONSTANT_AS u32a *s_td0 = td0;
+  CONSTANT_AS u32a *s_td1 = td1;
+  CONSTANT_AS u32a *s_td2 = td2;
+  CONSTANT_AS u32a *s_td3 = td3;
+  CONSTANT_AS u32a *s_td4 = td4;
+
+  CONSTANT_AS u32a *s_te0 = te0;
+  CONSTANT_AS u32a *s_te1 = te1;
+  CONSTANT_AS u32a *s_te2 = te2;
+  CONSTANT_AS u32a *s_te3 = te3;
+  CONSTANT_AS u32a *s_te4 = te4;
+  #endif
+
+  const u32 a = h32_from_64_S (tmps[gid].out[0]);
+  const u32 b = l32_from_64_S (tmps[gid].out[0]);
+  const u32 c = h32_from_64_S (tmps[gid].out[1]);
+  const u32 d = l32_from_64_S (tmps[gid].out[1]);
+  const u32 e = h32_from_64_S (tmps[gid].out[2]);
+  const u32 f = l32_from_64_S (tmps[gid].out[2]);
+  const u32 g = h32_from_64_S (tmps[gid].out[3]);
+  const u32 h = l32_from_64_S (tmps[gid].out[3]);
+
+  const u32 key[8] = { a,b,c,d,e,f,g,h };
+
+  u32 iv[4] = { 0 };
+  u32 res[64];
+  u32 ks[60];
+
+  AES256_set_decrypt_key (ks, key, s_te0, s_te1, s_te2, s_te3, s_td0, s_td1, s_td2, s_td3);
+
+  for (u32 i = 0; i < 64; i += 4)
+  {
+    u32 data[4];
+
+    data[0] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 0];
+    data[1] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 1];
+    data[2] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 2];
+    data[3] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 3];
+
+    u32 out[4];
+
+    aes256_decrypt (ks, data, out, s_td0, s_td1, s_td2, s_td3, s_td4);
+
+    res[i + 0] = hc_swap32_S (out[0] ^ iv[0]);
+    res[i + 1] = hc_swap32_S (out[1] ^ iv[1]);
+    res[i + 2] = hc_swap32_S (out[2] ^ iv[2]);
+    res[i + 3] = hc_swap32_S (out[3] ^ iv[3]);
+
+    iv[0] = data[0];
+    iv[1] = data[1];
+    iv[2] = data[2];
+    iv[3] = data[3];
+  }
+
+  u8 counter = 0;
```

Comment:
we normally do not use `u8` in the kernel code except if really, really required ("char" manipulation without any u32 and therefore "faster" alternative).
I think there is no reason for using `u8` here, you even did use `u32` in the loop above, so we should be consistent and just use `u32` for the `counter` variable and the `i` variable in the loop below etc. thx